### PR TITLE
Add aurora checker to SMS: Alex

### LINKED ISSUE #231: Add an SMS Alert for Auroras
As a user interested in auroras, I want to receive SMS alerts when auroras are visible near my location, so that I can be notified in real time and not miss out on witnessing the event.

- Add an SMS checker for any aurora alerts

### PR DESCRIPTION
**Pull request by AlexOBarnes**
This closes #231 

What I have done:
- Written tests for Aurora sms checker
- Implemented asynchronous calls for both email and sms function
- Implemented aurora check pipeline

### CODE REVIEW

File: sms-checker/checker.py [hunk 1]
```diff
@@ -9,7 +9,7 @@ def get_connection():
     return connect(f"""dbname={ENV["DB_NAME"]} user={ENV["DB_USER"]}
                  host={ENV["DB_HOST"]} password={ENV["DB_PASSWORD"]} port={ENV["DB_PORT"]}""")
 
-def get_subscribers() ->list[dict]:
+def get_subscribers_bodies() ->list[dict]:
```
Comment (olliejth) [inline sms-checker/checker.py:12]: Better name

File: sms-checker/checker.py [hunk 2]
```diff
@@ -37,8 +37,51 @@ def get_subscribers() ->list[dict]:
     return [{'user':sub[0],'phone':sub[1],'email':sub[2],
              'body':sub[3],'time':sub[4]} for sub in subscribers]
 
+def get_aurora_regions():
+    '''Returns a list of region ids for a given '''
+    query = '''SELECT ac.colour,ac.meaning FROM aurora_alert
+    JOIN aurora_colour as ac USING (aurora_colour_id)
+    ORDER BY alert_time DESC
+    LIMIT 1'''
```
Comment (olliejth) [inline sms-checker/checker.py:45]: Could be a bit more spaced out, quite crowded when written this way.

File: sms-checker/conftest.py [hunk 3]
```diff
@@ -13,6 +13,15 @@ def valid_clean_data():
              'body': 'Mars', 'time': dt(2024, 10, 1, 15, 30)}]
 
 
+@pytest.fixture
+def valid_clean_aurora_data():
+    '''Returns the expected data shape after cleaning'''
+    return [{'user': 'davidjohnson', 'phone': None,
+             'email': 'davidjohnson@example.com'},
```
Comment (olliejth) [inline sms-checker/conftest.py:20]: Efficient use of return.

File: sms-checker/message.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 from dotenv import load_dotenv
 from boto3 import client
 
-def send_sms(subscribers: list[dict]) -> None:
+def send_sms_for_bodies(subscribers: list[dict]) -> None:
```
Comment (olliejth) [inline sms-checker/message.py:7]: Better name.

File: sms-checker/message.py [hunk 5]
```diff
@@ -56,6 +56,52 @@ def send_email(subscribers: list[dict]) -> None:
             logging.warning('Invalid data in: %s',sub)
 
 
+def send_aurora_sms(sms, number, message):
```
Comment (olliejth) [inline sms-checker/message.py:59]: Add type hints.

**Review by olliejth — CHANGES_REQUESTED** [review]
Type hints requested in message.py.

File: sms-checker/message.py (hunk 5 above)
Comment (AlexOBarnes) [reply sms-checker/message.py:59]:
```suggestion
def send_aurora_sms(sms, number: str, message: str) -> None:
```

File: sms-checker/checker.py [hunk 6]
```diff
@@ -37,8 +37,51 @@ def get_subscribers() ->list[dict]:
...
+    JOIN region as r USING(region_id)
+    WHERE f.cloud_coverage_percent < 50 AND
+    f.visibility_m > 250'''
+
+    colour = get_aurora_regions()
+    logging.info('Current aurora alert level: %s',colour[0])
+    if colour[0] == 'Green':
+        return None,None
+    if colour[0] == 'Yellow':
+        query += 'AND r.region_id in [1,2,4,5,6]'
```
Comment (CalebMathers) [inline sms-checker/checker.py:72]: What about amber and red?

File: sms-checker/message.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def send_sms(subscribers: list[dict]) -> None:
             logging.warning('Invalid data in: %s', sub)
 
 
-def send_email(subscribers: list[dict]) -> None:
+def send_email_for_bodies(subscribers: list[dict]) -> None:
```
Comment (CalebMathers) [inline sms-checker/message.py:30]: Good descriptive function name

File: sms-checker/notification_lambda.py [hunk 8]
```diff
@@ -1,20 +1,68 @@
...
+
+
+async def alerts_for_bodies() -> None:
+    '''Asynchronous function that alerts users about celestial bodies'''
+    subscribers = get_subscribers_bodies()
+    if subscribers:
+        logging.info('Subscribers found: %s', len(subscribers))
+        email_list, phone_list = process_users(subscribers)
+        logging.info('Emails to be sent: %s', len(email_list))
+        logging.info('Texts to be sent: %s', len(phone_list))
```
Comment (CalebMathers) [inline sms-checker/notification_lambda.py:23]: Great use of logging so we know how many texts/emails we are sending out

File: sms-checker/notification_lambda.py [hunk 9]
```diff
@@ -1,20 +1,68 @@
...
+            tasks.append(send_email_for_bodies(email_list))
+        if phone_list:
+            tasks.append(send_sms_for_bodies(phone_list))
+        if tasks:
+            await asyncio.gather(*tasks)
+    else:
+        logging.info('No subscribers found.')
+
+
+async def alerts_for_auroras() -> None:
```
Comment (CalebMathers) [inline sms-checker/notification_lambda.py:35]: Good use of async to make things more efficient

File: sms-checker/notification_lambda.py [hunk 10]
```diff
@@ -1,20 +1,68 @@
...
+        logging.info('No subscribers found.')
+
+
+async def run_notifications():
+    '''Runs asynchronous functions'''
+    await asyncio.gather(alerts_for_bodies(),
+                         alerts_for_auroras())
+
 
 def lambda_handler(event,context) -> None:
```
Comment (olliejth) [inline sms-checker/notification_lambda.py:60]: Can add default values of 'None' in the future for lambda handler functions.

**Review by olliejth — APPROVED** [review]
Valid changes made.

File: sms-checker/test_checker.py
Comment (CalebMathers) [inline sms-checker/test_checker.py:1]: Solid tests

File: sms-checker/test_message.py [hunk 11]
```diff
@@ -1,63 +1,134 @@
...
         '''Tests that an empty list does not result in a call'''
         mock_ses = mock.MagicMock()
         mock_get_client.return_value = mock_ses
         subscribers = []
-        send_email(subscribers)
+        send_email_for_bodies(subscribers)
         mock_ses.send_email.assert_not_called()
+
+
+class TestAssignContent:
```
Comment (CalebMathers) [inline sms-checker/test_message.py:68]: Good to test for each eventuality

**Review by CalebMathers — APPROVED** [review]

**Pull request merged**